Initialize CBCentralManager when listening for availability changes

fotiDim opened the pull request:
This PR addresses https://github.com/woodemi/quick.flutter/issues/42.

Sunbreak commented on packages/quick_blue/darwin/QuickBlueDarwin.swift:
```diff
@@ -67,6 +67,7 @@ public class QuickBlueDarwin: NSObject, FlutterPlugin {
   private var discoveredPeripherals = Dictionary<String, CBPeripheral>()
 
   private var scanResultSink: FlutterEventSink?
+  private var availabilityChangeSink: FlutterEventSink?
```
We'd better declare `availabilityChangeSink` before `scanResultSink`

Sunbreak commented on packages/quick_blue/darwin/QuickBlueDarwin.swift:
```diff
@@ -220,6 +221,11 @@ extension QuickBlueDarwin: FlutterStreamHandler {
     print("QuickBlueDarwin onListenWithArguments: \(name)")
     if name == "scanResult" {
       scanResultSink = events
+    } else if name == "availabilityChange" {
```
availabilityChange first

Sunbreak commented on packages/quick_blue/darwin/QuickBlueDarwin.swift:
```diff
@@ -231,6 +237,8 @@ extension QuickBlueDarwin: FlutterStreamHandler {
     print("QuickBlueDarwin onCancelWithArguments: \(name)")
     if name == "scanResult" {
       scanResultSink = nil
+    } else if name == "availabilityChange" {
```
availabilityChange first

Sunbreak commented on packages/quick_blue/example/lib/main.dart:
```diff
@@ -53,11 +53,27 @@ class _MyAppState extends State<MyApp> {
         ),
         body: Column(
           children: [
+            // StreamBuilder<AvailabilityState>(  // Uncomment on iOS and macOS
```
Use [collection if](https://dart.dev/guides/language/language-tour#collection-operators)

```dart
if (Platform.isIOs && Platform.isMacOS)
  StreamBuild<AvailabilityState>
```

Sunbreak commented on packages/quick_blue/example/lib/main.dart:
```diff
@@ -53,11 +53,27 @@ class _MyAppState extends State<MyApp> {
+            //   stream: QuickBlue.availabilityChangeStream,
+            //   builder: (context, snapshot) {
+            //     String availability;
+            //     switch (snapshot.data) {
```
```
extension AvailabilityStateExt on AvailabilityState {
  String toString() {
    switch() {
      case xx: return 'xx';
    }
  }
}
```

fotiDim replied:
I put it in the model and overrode toString() since this cannot be done in an extension.

Sunbreak commented on packages/quick_blue/lib/quick_blue.dart:
```diff
@@ -39,6 +39,11 @@ class QuickBlue {
       .map((item) => BlueScanResult.fromMap(item));
   }
 
+  static Stream<AvailabilityState> get availabilityChangeStream {
```
Move `get availabilityChangeStream` above `setAvailabilityHandler` and remove `setAvailabilityHandler`

Sunbreak replied:
This seems not revolved

Sunbreak commented on packages/quick_blue/darwin/QuickBlueDarwin.swift:
```diff
@@ -220,6 +221,11 @@ extension QuickBlueDarwin: FlutterStreamHandler {
       scanResultSink = events
+    } else if name == "availabilityChange" {
+      availabilityChangeSink = events
+      availabilityChangeSink?([
```
We'd better leave a comment here `// FIXME https://github.com/woodemi/quick.flutter/issues/42`

fotiDim replied:
It is fixed already. Why do we need the comment?

Sunbreak replied:
We usually dont fire an event as soon as `onListen`. A `// HACK` may indicate why the code is here

fotiDim replied:
It is not really a hack. It actually serves a double purpose:
- Initializes CBCentralManager
- Returns state when hot restarting otherwise CBCentralManager is already initialized and no state is returned.

I added a comment to indicate this.

Sunbreak commented on packages/quick_blue/lib/src/method_channel_quick_blue.dart:
```diff
@@ -45,10 +45,14 @@ class MethodChannelQuickBlue extends QuickBluePlatform {
   }
 
   final Stream<dynamic> _scanResultStream = _eventScanResult.receiveBroadcastStream({'name': 'scanResult'});
+  final Stream<dynamic> _availabilityChangeStream = _eventScanResult.receiveBroadcastStream({'name': 'availabilityChange'});
```
Move `_availabilityChangeStream` above `startScan` and below `isBluetoothAvailable`

Sunbreak commented on packages/quick_blue/lib/src/method_channel_quick_blue.dart:
```diff
@@ -45,10 +45,14 @@ class MethodChannelQuickBlue extends QuickBluePlatform {
   Stream<dynamic> get scanResultStream => _scanResultStream;
 
+  @override
+  Stream<dynamic> get availabilityChangeStream => _availabilityChangeStream;
```
Move `get availabilityChangeStream` above `startScan` and below `_availabilityChangeStream`

Sunbreak commented on packages/quick_blue/lib/src/quick_blue_linux.dart:
```diff
@@ -187,6 +187,10 @@ class QuickBlueLinux extends QuickBluePlatform {
+  
+  @override
+  // TODO: implement availabilityChangeStream
+  Stream get availabilityChangeStream => throw UnimplementedError();
```
Move `_availabilityChangeStream` above `startScan` and below `isBluetoothAvailable`

Sunbreak commented on packages/quick_blue/lib/src/quick_blue_platform_interface.dart:
```diff
@@ -45,6 +45,8 @@ abstract class QuickBluePlatform extends PlatformInterface {
 
   Stream<dynamic> get scanResultStream;
 
+  Stream<dynamic> get availabilityChangeStream;
```
Move `get availabilityChangeStream` above `startScan` and below `isBluetoothAvailable`

Sunbreak commented on packages/quick_blue/lib/src/models.dart:
```diff
@@ -28,6 +28,34 @@ class AvailabilityState {
+    String availability;
+    switch (this) {
+      case AvailabilityState.unknown:
+        availability = 'unknown';
```
return 'unknown';

fotiDim commented:
@Sunbreak anything pending on this PR? We would like to implement this on additional platforms.

Sunbreak commented:
> <img width="538" alt="Screen Shot 2022-07-07 at 9 57 47 AM" src="https://user-images.githubusercontent.com/7928961/177673706-04c6bffd-d179-4890-a40f-935dd76eb5ae.png">

An unresolved comment in the hidden conversation

> <img width="830" alt="Screen Shot 2022-07-07 at 9 58 18 AM" src="https://user-images.githubusercontent.com/7928961/177673753-9e54c390-a275-4f84-ac18-15e366f3c50c.png">

fotiDim commented:
@Sunbreak sorry I missed that. Fixed now. I also made some other improvements:
- Removed all traces of setAvailabilityHandler
- created a separate event channel
- updated the example app

Please have another look.

Sunbreak commented on packages/quick_blue/darwin/QuickBlueDarwin.swift:
```diff
@@ -55,17 +55,20 @@ public class QuickBlueDarwin: NSObject, FlutterPlugin {
     #endif
     let method = FlutterMethodChannel(name: "quick_blue/method", binaryMessenger: messenger)
     let eventScanResult = FlutterEventChannel(name: "quick_blue/event.scanResult", binaryMessenger: messenger)
+    let eventAvailabilityChange = FlutterEventChannel(name: "quick_blue/event.availabilityChange", binaryMessenger: messenger)
```
Move `eventAvailabilityChange` before `eventScanResult`

Sunbreak commented on packages/quick_blue/lib/src/models.dart:
```diff
@@ -28,6 +28,26 @@ class AvailabilityState {
+      case AvailabilityState.poweredOn:
+        return 'poweredOn';
+      default:
+        return '...';
```
throw ArguementException();

Sunbreak commented on packages/quick_blue/darwin/QuickBlueDarwin.swift:
```diff
@@ -218,7 +219,10 @@ extension QuickBlueDarwin: FlutterStreamHandler {
-    if name == "scanResult" {
+    if name == "availabilityChange" {
+      availabilityChangeSink = events
+      availabilityChangeSink?(manager.state.rawValue)
```
Missing `// HACK` comment

fotiDim commented:
done
